File: infrastructure/deploy-schema.py
# AppSync Schema Deployment Script
# Uses AWS CLI to deploy GraphQL schema

import json
import boto3
import time
import sys
from pathlib import Path

# Configuration
SCRIPT_DIR = Path(__file__).parent
CONFIG_FILE = SCRIPT_DIR / "appsync-config.json"
SCHEMA_FILE = SCRIPT_DIR / "schema.graphql"

# Load configuration
try:
    with open(CONFIG_FILE, 'r') as f:
        config = json.load(f)
        API_ID = config['ApiId']
        REGION = config['Region']
except FileNotFoundError:
    print("❌ Error: appsync-config.json not found")
    sys.exit(1)
except KeyError as e:
    print(f"❌ Error: Missing key in config: {e}")
    sys.exit(1)

print("🚀 Deploying GraphQL Schema to AppSync")
print(f"API ID: {API_ID}")
print(f"Region: {REGION}")
print()

# Read schema
try:
    with open(SCHEMA_FILE, 'r', encoding='utf-8') as f:
        schema_content = f.read()
except FileNotFoundError:
    print(f"❌ Error: {SCHEMA_FILE} not found")
    sys.exit(1)

print(f"📄 Read schema ({len(schema_content)} characters)")

# Initialize AWS AppSync client
client = boto3.client('appsync', region_name=REGION)

try:
    print("⬆️  Uploading schema to AppSync...")
    
    # Start schema creation
    response = client.start_schema_creation(
        apiId=API_ID,
        definition=schema_content.encode('utf-8')
    )
    
    print(f"✅ Schema upload started (status: {response['status']})")
    
    # Poll for completion
    print("⏳ Waiting for deployment to complete...")
    max_attempts = 60
    attempt = 0
    
    while attempt < max_attempts:
        time.sleep(2)
        attempt += 1
        
        status_response = client.get_schema_creation_status(apiId=API_ID)
        status = status_response['status']
        
        print(f"  Status: {status} (attempt {attempt}/{max_attempts})")
        
        if status == 'SUCCESS':
            print()
            print("✅ Schema deployed successfully!")
            break
        elif status == 'FAILED':
            details = status_response.get('details', 'No details provided')
            print()
            print(f"❌ Schema deployment failed: {details}")
            sys.exit(1)
        elif status in ['PROCESSING', 'PENDING']:
            # Continue waiting
            pass
        else:
            print(f"⚠️  Unknown status: {status}")
    
    if status != 'SUCCESS':
        print()
        print(f"❌ Deployment timed out after {max_attempts} attempts")
        sys.exit(1)
    
    print()
    print("═" * 50)
    print("🎉 Schema Deployment Complete!")
    print("═" * 50)
    print()
    print("Next steps:")
    print("1. Deploy resolvers: .\\deploy-resolvers-only.ps1")
    print("2. Test queries in AppSync console")
    print("3. Refresh your web app")
    print()
    
except Exception as e:
    print(f"❌ Error: {e}")
    sys.exit(1)

File: infrastructure/deploy-schema.js
// AppSync Schema Deployment Script
// Uses AWS SDK for JavaScript to deploy GraphQL schema

const fs = require('fs');
const path = require('path');
const { AppSyncClient, StartSchemaCreationCommand, GetSchemaCreationStatusCommand } = require('@aws-sdk/client-appsync');

// Configuration
const configPath = path.join(__dirname, 'appsync-config.json');
const schemaPath = path.join(__dirname, 'schema.graphql');

// Load configuration
let config;
try {
    config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
} catch (error) {
    console.error('❌ Error: Failed to load appsync-config.json');
    process.exit(1);
}

const { ApiId, Region } = config;

console.log('🚀 Deploying GraphQL Schema to AppSync');
console.log(`API ID: ${ApiId}`);
console.log(`Region: ${Region}`);
console.log();

// Read schema
let schemaContent;
try {
    schemaContent = fs.readFileSync(schemaPath, 'utf8');
    console.log(`📄 Read schema (${schemaContent.length} characters)`);
} catch (error) {
    console.error('❌ Error: Failed to read schema.graphql');
    process.exit(1);
}

// Initialize AWS AppSync client
const client = new AppSyncClient({ region: Region });

async function deploySchema() {
    try {
        console.log('⬆️  Uploading schema to AppSync...');
        
        // Start schema creation
        const startCommand = new StartSchemaCreationCommand({
            apiId: ApiId,
            definition: Buffer.from(schemaContent, 'utf8')
        });
        
        const startResponse = await client.send(startCommand);
        console.log(`✅ Schema upload started (status: ${startResponse.status})`);
        
        // Poll for completion
        console.log('⏳ Waiting for deployment to complete...');
        const maxAttempts = 60;
        let attempt = 0;
        let deployed = false;
        
        while (attempt < maxAttempts && !deployed) {
            await new Promise(resolve => setTimeout(resolve, 2000));
            attempt++;
            
            const statusCommand = new GetSchemaCreationStatusCommand({ apiId: ApiId });
            const statusResponse = await client.send(statusCommand);
            const status = statusResponse.status;
            
            console.log(`  Status: ${status} (attempt ${attempt}/${maxAttempts})`);
            
            if (status === 'SUCCESS') {
                deployed = true;
                console.log();
                console.log('✅ Schema deployed successfully!');
            } else if (status === 'FAILED') {
                const details = statusResponse.details || 'No details provided';
                console.log();
                console.error(`❌ Schema deployment failed: ${details}`);
                process.exit(1);
            } else if (status === 'PROCESSING' || status === 'PENDING') {
                // Continue waiting
            } else {
                console.warn(`⚠️  Unknown status: ${status}`);
            }
        }
        
        if (!deployed) {
            console.log();
            console.error(`❌ Deployment timed out after ${maxAttempts} attempts`);
            process.exit(1);
        }
        
        console.log();
        console.log('═'.repeat(50));
        console.log('🎉 Schema Deployment Complete!');
        console.log('═'.repeat(50));
        console.log();
        console.log('Next steps:');
        console.log('1. Deploy resolvers: .\\deploy-resolvers-only.ps1');
        console.log('2. Test queries in AppSync console');
        console.log('3. Refresh your web app');
        console.log();
        
    } catch (error) {
        console.error(`❌ Error: ${error.message}`);
        if (error.$metadata) {
            console.error('Details:', error.$metadata);
        }
        process.exit(1);
    }
}

deploySchema();
